src/client.ts

```diff
@@ -1,149 +1,224 @@
+    all the step functions. 
+  */
+  private static accessToken: string;
+
+  constructor(readonly config: CalculatedIntegrationConfig) {
+    this.accountId = config.accountId;
+    this.clientId = config.clientId;
+    this.clientSecret = config.clientSecret;
+  }
 
-  public async verifyAuthentication(): Promise<void> {
-    // TODO make the most light-weight request possible to validate
-    // authentication works with the provided credentials, throw an err if
-    // authentication fails
-    const request = new Promise((resolve, reject) => {
-      http.get(
-        {
-          hostname: 'localhost',
-          port: 443,
-          path: '/api/v1/some/endpoint?limit=1',
-          agent: false,
-          timeout: 10,
-        },
-        (res) => {
-          if (res.statusCode !== 200) {
-            reject(new Error('Provider authentication failed'));
-          } else {
-            resolve();
-          }
-        },
+  private withBaseUri(path: string): string {
+    return `https://api.malwarebytes.com/${path}`;
+  }
+
+  private async request(
+    uri: string,
+    method: 'GET' | 'POST' | 'HEAD' = 'GET',
+    body?: any,
+  ): Promise<Response> {
+    if (!APIClient.accessToken) {
+      await this.getAuthenticationToken();
+    }
+
+    return await fetch(uri, {
+      method,
+      headers: {
+        Authorization: `Bearer ${APIClient.accessToken}`,
+        AccountId: this.accountId,
+        'Content-Type': 'application/json',
+      },
+      body: body || null,
+    });
+  }
+
+  private async makePaginatedRequest<K extends string, V>(
+    uri: string,
+    method: 'GET' | 'POST' | 'HEAD' = 'GET',
+    resource: K,
+    body?: any,
+  ): Promise<V[]> {
+    const results: V[] = [];
+    let nextCursor = '';
+
+    do {
+      const response = await this.request(
+        uri,
+        method,
+        JSON.stringify({
+          ...body,
+          page_size: 10,
+          next_cursor: nextCursor,
+        }),
       );
+
+      const responseJson: PaginatedDataResponse<K, V[]> = await response.json();
+
+      if (responseJson) {
+        nextCursor = responseJson.next_cursor;
+        results.push(...responseJson[resource]);
+      }
+    } while (nextCursor && nextCursor.length > 0);
+
+    return results;
+  }
+
+  private async getAuthenticationToken() {
+    const getAuthTokenEndpoint = this.withBaseUri('oauth2/token');
+    const response = await fetch(getAuthTokenEndpoint, {
+      method: 'POST',
+      headers: {
+        Authorization: `Basic ${Buffer.from(
+          `${this.clientId}:${this.clientSecret}`,
+        ).toString('base64')}`,
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      body: qs.stringify({
+        scope: 'read',
+        grant_type: 'client_credentials',
+      }),
     });
 
     try {
-      await request;
-    } catch (err) {
+      const body = await response.json();
+      APIClient.accessToken = body.access_token;
```

Comment:
Nice thought, setting `APIClient.accessToken` instead of `this.accessToken`, so this can be shared across steps.

Does this accessToken have a TTL that we need to be concerned about? Will it need to be refreshed for any long-running integrations?

Reply:
Actually, allow me to amend my statement. While it does offer the benefit of sharing the `accessToken` across steps, it is in contrast with the paradigm we use of not mutating state, which can lead to some confusing bugs in the future. Can you change this to `this.accessToken`?
```suggestion
      this.accessToken = body.access_token;
```

Reply:
Sounds good! We've changed it from static to instance.